File: cacheable/cacheable.js
import fetch from "node-fetch"

export class CacheableFetch {
  cache = new Map()

  async get(endpoint){
    if(this.cache.has(endpoint)){
      return {
        data: this.cache.get(endpoint),
        status: 'cached'
      }
    }

    const result = await fetch(endpoint)
    const data = await result.json()

    this.cache.set(endpoint, data)

    return {
      data: data,
      status: 'first time'
    }
  }

  invalidate(path){
    this.cache.delete(path)
  }
}

File: cacheable-dart/cacheable.dart
import 'package:dio/dio.dart';

class Cacheable {
  Map<String, Object> cache = {};

  Future get({ url }) async {
    if(this.cache.containsKey(url)){
      return {
        'data': this.cache[url],
        'status': 'cached'
      };
    }

    var result = await Dio().get(url);
    this.cache[url] = result.data;

    return {
      'data': result,
      'status': 'first time'
    };
  }

  invalidate({ url }) async {
    this.cache.remove(url);
  }
}
